edgePy/DGEList.py

```diff
@@ -292,17 +292,40 @@ def rpkm(
         # self = self.cpm(log=log, prior_count=prior_count)
 
     def tpm(
-        self, transcripts: Mapping, log: bool = False, prior_count: float = PRIOR_COUNT
-    ) -> None:
-        """Return the DGEList normalized to reads per kilobase of transcript
-        length.
+        self,
+        gene_lengths: np.ndarray,
+        log: bool = False,
+        prior_count: float = PRIOR_COUNT,
+        mean_fragment_lengths: np.ndarray = None,
+    ) -> np.ndarray:
+        """Normalize the DGEList to transcripts per million
+
+        Args:
+            gene_lengths: vector of length for each gene in counts table
+            log: store log outputs
+            prior_count:
+            mean_fragment_lengths: vector of mean fragment length for each sample in counts table (optional)
 
         """
-        raise NotImplementedError
 
-        # TODO: Implement here
+        # based on https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ and
+        # https://gist.github.com/slowkow/c6ab0348747f86e2748b
+        # how many counts per base
 
-        # self = self.cpm(log=log, prior_count=prior_count)
+        # compute effective length not allowing negative lengths
+        if mean_fragment_lengths:
+            effective_lengths = (gene_lengths[:, None] - mean_fragment_lengths[None, :]).clip(
+                min=1
+            )
+        else:
+            effective_lengths = gene_lengths[:, None]
+
+        base_counts = self.counts / effective_lengths
+
+        self.counts = 10 ** 6 * base_counts / np.sum(base_counts, axis=0)[None, :]
```

Comment:
Order of operations is clear but just to be safe I would group with `()`.


```python
self.counts = (10 ** 6) * base_counts / np.sum(base_counts, axis=0)[None, :]
```

Or better yet, drop the slicing and use `e`:

```python
self.counts = 10e6 * base_counts / np.sum(base_counts, axis=0)
```